File: src/Striver_SDE_Sheet/Stack_and_Queue/LFUCache/LFUCache.cpp
#include <bits/stdc++.h>
#include <list>
#include <unordered_map>
using namespace std;

class LFUCache {
    
    inline static const int NO_SUCH_KEY = -1;
    size_t capacity;
    unordered_map<int, int> keyToValue;
    unordered_map<int, int> keyToFrequency;
    unordered_map<int, list<int>> frequencyToKey;

public:
    LFUCache(int capacity) {
        this->capacity = capacity;
    }

    int get(int key) {
        if (keyToValue.find(key) == keyToValue.end()) {
            return NO_SUCH_KEY;
        }

        int previousFrequency = keyToFrequency[key];
        keyToFrequency[key] = previousFrequency + 1;
        updateMapFrequencyToKey(previousFrequency, key);

        return keyToValue[key];
    }

    void put(int key, int value) {
        if (capacity == 0) {
            return;
        }

        if (keyToValue.size() == capacity && keyToValue.find(key) == keyToValue.end()) {
            int leastFrequentlyUsed = INT_MAX;
            for (const auto& frequency : frequencyToKey) {
                leastFrequentlyUsed = min(leastFrequentlyUsed, frequency.first);
            }
            removeLeastRecentlyUsedFromLeastFrequentlyUsed(leastFrequentlyUsed);
        }

        int previousFrequency = keyToFrequency.find(key) != keyToFrequency.end() ? keyToFrequency[key] : 0;
        keyToValue[key] = value;
        keyToFrequency[key] = previousFrequency + 1;
        updateMapFrequencyToKey(previousFrequency, key);
    }

private:
    void updateMapFrequencyToKey(int previousFrequency, int key) {
        if (frequencyToKey.find(previousFrequency + 1) == frequencyToKey.end()) {
            frequencyToKey[previousFrequency + 1] = list<int>();
        }
        frequencyToKey[previousFrequency + 1].push_back(key);

        if (frequencyToKey.find(previousFrequency) != frequencyToKey.end()) {
            frequencyToKey[previousFrequency].remove(key);
            if (frequencyToKey[previousFrequency].empty()) {
                frequencyToKey.erase(previousFrequency);
            }
        }
    }

    void removeLeastRecentlyUsedFromLeastFrequentlyUsed(int leastFrequentlyUsed) {
        int keyToRemove = frequencyToKey[leastFrequentlyUsed].front();
        frequencyToKey[leastFrequentlyUsed].pop_front();

        keyToValue.erase(keyToRemove);
        keyToFrequency.erase(keyToRemove);

        if (frequencyToKey[leastFrequentlyUsed].empty()) {
            frequencyToKey.erase(leastFrequentlyUsed);
        }
    }
};
/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: src/Striver_SDE_Sheet/Stack_and_Queue/LFUCache/LFUCache.java
package Striver_SDE_Sheet.Stack_and_Queue.LFUCache;

import java.util.*;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;

public class LfuCache {
    private final Map<Integer, Integer> values = new HashMap<>();
    private final Map<Integer, Integer> counts = new HashMap<>();
    private final Map<Integer, LinkedHashSet<Integer>> frequency = new HashMap<>();

    private int capacity = -1;
    private int minFrequency = -1;

    public LfuCache(int capacity) {
        this.capacity = capacity;
    }

    public int get(int key) {
        if (!values.containsKey(key)) {
            return -1;
        }

        int keyCount = counts.get(key);
        counts.put(key, keyCount + 1);

        frequency.get(keyCount).remove(key);
        if (keyCount == minFrequency && frequency.get(keyCount).isEmpty()) {
            minFrequency++;
        }

        if (!frequency.containsKey(keyCount + 1)) {
            frequency.put(keyCount + 1, new LinkedHashSet<>());
        }
        frequency.get(keyCount + 1).add(key);

        return values.get(key);
    }

    public void put(int key, int value) {
        if (capacity <= 0) {
            return;
        }

        if (values.containsKey(key)) {
            values.put(key, value);
            get(key);
            return;
        }

        if (values.size() >= capacity) {
            int evicted = frequency.get(minFrequency).iterator().next();
            frequency.get(minFrequency).remove(evicted);
            values.remove(evicted);
            counts.remove(evicted);
        }

        values.put(key, value);
        counts.put(key, 1);
        minFrequency = 1;
        if (!frequency.containsKey(1)) {
            frequency.put(1, new LinkedHashSet<>());
        }
        frequency.get(1).add(key);
    }
}
/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */